### DISCUSSION TITLE
3. 데이터베이스의 종류와 엔진

### DISCUSSION THREAD

**Discussion by jindream6128** [24tngus/CS_STUDY#16] [Database]
## 데이터 베이스는 크게 RDBMS / NoSQL로 나뉜다.

1. ### 데이터 모델링:
    - RDBMS: RDBMS는 테이블 기반 데이터 모델을 사용합니다. `데이터는 테이블에 행과 열`로 저장되며, 정해진 스키마를 따라야 합니다. 각 테이블은 기본 키(primary key)에 의해 고유하게 식별됩니다. 데이터 간의 관계를 정의하고 관리하는 데 강력하며, 관계형 대수를 사용하여 데이터를 조작합니다.
    - NoSQL (MongoDB): NoSQL은 다양한 데이터 모델을 제공합니다. MongoDB는 `Document-Oriented `데이터베이스로, JSON 형식의 도큐먼트를 사용하여 데이터를 저장합니다. 각 도큐먼트는 유연한 스키마를 가지므로 필드의 추가 및 삭제가 쉽습니다. 데이터 간의 관계를 강제하지 않고, 중첩된 구조를 허용하여 복잡한 데이터 구조를 저장할 수 있습니다.
    
2. ### 스키마:
    - RDBMS: RDBMS는 `정적 스키마`를 가집니다. 데이터를 저장하기 전에 미리 정의된 스키마를 생성해야 합니다. 이로 인해 데이터 구조 변경이 어려울 수 있습니다.
    - NoSQL (MongoDB): NoSQL은 `동적 스키마`를 가지고 있습니다. 데이터를 저장할 때마다 필요에 따라 스키마를 변경할 수 있습니다. 이는 데이터 모델링을 더 유연하게 하며, 신속한 개발과 변경에 유리합니다.
    
3. ### 확장성:
    - RDBMS: 전통적인 RDBMS는 `수직적 확장(Vertical Scaling)에 적합`합니다. 즉, 더 높은 처리량을 위해 단일 서버의 성능을 향상시키는 방식으로 확장할 수 있습니다. 하지만 한계에 도달하면 더 이상 확장이 어려울 수 있습니다.
    - NoSQL (MongoDB): NoSQL은 `수평적 확장(Horizontal Scaling)에 적합`합니다. 여러 서버에 데이터를 분산하여 처리량을 늘릴 수 있으며, 클라우드 기반 아키텍처에서 쉽게 확장할 수 있습니다.
    
4. ### 트랜잭션 처리:
    - RDBMS: RDBMS는 `ACID(원자성, 일관성, 고립성, 지속성) 트랜잭션`을 지원합니다. 이는 데이터의 안전한 처리를 보장하며, 트랜잭션 중 하나의 단계가 실패하면 롤백하여 데이터 일관성을 유지합니다.
    - NoSQL (MongoDB): NoSQL은 대부분의 경우 `BASE(Basically Available, Soft state, Eventually consistent)`를 따릅니다. ACID보다 유연한 모델을 가지며, 일관성보다 가용성과 성능을 우선시합니다. 따라서 데이터 일관성을 보장하는 데에는 좀 더 신경을 써야 할 수 있습니다.
    
5. ### 쿼리언어
    - RDBMS: RDBMS는 대부분 `SQL(Structured Query Language)을 사용하여 데이터를 조작`합니다.
    - NoSQL (MongoDB): MongoDB는 `MongoDB Query Language (MQL)을 사용하여 데이터를 조회하고 조작`합니다. MQL은 JSON과 유사한 문법을 가지고 있습니다.

=> 즉 RDBMS는 정형화된 데이터의 모델과, 강력한 트랜잭션 처리가 필요한 경우 적합.
 - RDBMS 에는 Mysql, Oracle, MSSQL 등이 존재하고 은행,금융, 웹어플리케이션 등 대부분이 DB는 RDBMS를 사용

=> NoSQL은 유연한 데이터 모델과 대량의 데이터를 처리해야 하는 경우 더 적합.
-  NoSql에는 Mongodb,Redis 등이 존재하고 크롤링이나 빅데이터 같은 다량의 데이터를 처리함에 있어 자주 사용.

### RDBMS 의 MYSQL에 사용되는 엔진

`InnoDB vs MyISAM`

`InnoDB` → 트랜잭션 처리가 필요하고, 대용량의 데이터를 다루는 부분에서 효율적

`MyISAM`→ 트랜잭션 처리가 필요없고, Read Only의 기능이 많은 서비스

`결론` => 데이터 변화가 많은 서비스의 경우 InnoDB 엔진을, SELECT(조회) 만 많은 서비스의 경우 MyISAM을 사용하고, 테이블을 만들때 데이터베이스의 엔진을 결정할 수 있다.

`주의점` => InnoDB와 MyISAM 엔진을 각각 사용하는 테이블을 서로 조인할때에 조심해야 한다.

### NoSQL 의 Mongodb에 사용되는 엔진

`WiredTiger vs MMAPv1`

`WiredTiger(와이어드 타이거)` -> Mongodb 3.0 부터 기본 엔진으로서, 트랜잭션 지원과 ACID 특성을 가짐으로써 데이터의 일관성과 안정성을 보당하고 있다. 또한 압축 기능을 통해 더욱 저장공간을 절약하고, MVCC(Multi Version Concurrency Control)을 기반으로 하며, 동시에 여러 트랜잭션을 처리할수 있어 동시성이 좋다.

`MMAPv1` -> 이전 버전의 기본 엔진으로써, 일부분에 있어서는 와이어드 타이거 엔진보다 빠를수도 있지만, 비교적 제한적이며 더 이상의 사용을 지양 한다.

`결론` => 대부분 WiredTiger 엔진이 많은 기능을 가지고있고, 높은 성능을 가지고있어 해당 엔진을 사용하고 있고, MMAPv1의 경우 옛날에 사용되던 엔진으로써 사용을 지양한다고 한다.

### MySQL 5.7* version과 8.*버전의 차이

1. JSON 지원 및 작업 향상:
MySQL 8.0에서는 JSON 데이터 유형의 지원과 관련하여 다양한 개선이 이루어졌습니다. 

2. 성능 최최적화
InnoDB 스토리지 엔진에 대한 여러 성능 향상으로 성능이 최적화 되었습니다.

3. 인증 및 보안 강화
인증과 보안이 강화되었습니다. 비밀번호 정책, 플러그형 인증 등 SSL TLS 지원 등을 포함하여 보안 관련 개선 사항이 많이 추가되었습니다.

4. 캐싱기능
 - 5.7버전의 경우 Query Cache(쿼리캐시) 기능을 지원하여, 쿼리와 그 결과를 메모리에 저장하여 동일한 쿼리에 대한 결괄르 더 빠르게 제공하는 역할을 하는 기능 입니다. 하지만 쿼리 캐시는 특정 시나리오에서 성능 문제를 일으킬 수 있고, 멀티스레드 환경에서는 더이상 권장되지 않기 때문에. 8.* 버전에서 삭제 되었습니다.
 - 8.* 버전에서는 쿼리 캐시기능이 제거되었지만, InnoDB 버퍼 풀과 새로운 쿼리 실행 계획 옵티마이저가 데이터 엑세스 및 쿼리 실행 계획 관리에 사용됩니다. -> InnoDB 스토리지 엔진의 최적화에 따른 결과

**Comment by 24tngus** [comment]
확인했습니다!

**Comment by berryberries** [comment]
확인했습니다.

**Comment by jindream6128** [comment]
멘토님 Comment)
빅데이터를 다룰 때 NoSQL을 사용하더라구요. (트위터에서도 사용하며 트윗이 Document라고 합니다) 보통 흔히 하는 프로젝트의 요구사항엔 RDBMS가 적합한 경우가 많습니다

**Comment by jindream6128** [comment]
저번에 수진님께서 피드백 해주신 MYSQL 5.** 버전과, 8.** 버전의 차이에 대해 조금 알아 봤는데! 그때 수진님이 말씀해주신게 뭔지 기억이 안나가지고,,
일단 가장 큰 차이는 8.** 부터 JSON 형식의 지원이 조금 더 많아져서 실무에서는 5.* 버전을 쓰는곳이 몇 없다고 하더라구요 !
5.7버전 이전에는 json을 문자열로 바꾸고 데이터를 저장하는 방식이었다면 이후에는 Json을 별도의 데이터타입으로 지원해준다고 합니다 !
또안 보안기능 강화 원래 스키마 자체의 비밀번호를 MYSQL의 테이블에서 쉽게 확인이 가능하였는데, 암호화되어 테이블에 저장되어 더욱 보안에 강화되었다고 합니다 !

**Reply by sujin-park0607** [reply]
MysQL에서 캐싱기능이 좋기는 한데 테이블을 수정하게되면 캐시까지 수정해야하는데 그때, Thread에 락을 건다고합니다! 근데 그게 심각하게 성능이 안좋아져서 8.0에서는 캐싱기능이 빠진거라고 하네요 ㅎㅎ 찾아봐주셔서 감사합니다 추가로 설명해주신 8.0을 쓰는 이유도 잘 배워갈게요!

**Comment by solo5star** [comment]
확인했습니다! 좀 딥하게 공부하는 백엔드 분은 Real MySQL 책 읽으시더라구요. DB를 더 공부해본다면 이 책을 한번 읽어보면 좋을 것 같습니다~

**Comment by jindream6128** [comment]
레퍼지토리 업로드 완료.
